blog/articles/event-driven-image-bigquery-processing-pipelines.md

```diff
@@ -0,0 +1,158 @@
+the following resized and watermarked image:
+
+![Beach with sunset](https://atamel.dev/img/2020/beach-400x400-watermark.jpeg)
+
+## BigQuery Processing Pipeline
+
+In this second pipeline, I built a schedule driven pipeline. In this pipeline, I query
+find the daily COVID-19 cases for a couple of countries. I use a public COVID-19
+dataset on BigQuery to get the data, generate some charts and send myself an
+email for each country once a day with those charts.
+
+### Architecture
+
+Here's the architecture of the pipeline.
+
+![BigQuery processing pipeline architecture](https://atamel.dev/img/2020/bigquery-processing-pipeline.png)
+
+1. Two `CloudSchedulerSources` are setup for two countries (United Kingdom and
+   Cyprus) to call the `QueryRunner` service once a day.
+2. `QueryRunner` service written in C#. It receives the scheduler event for both
+   countries, queries Covid-19 cases for the country using BigQuery's public
+   Covid-19 dataset and saves the result in a separate BigQuery table. Once
+   done, `QueryRunner` returns a custom `CloudEvent` of type
+   `dev.knative.samples.querycompleted`.
+3. `ChartCreator` service written in Python. It receives the `querycompleted`
+   event, creates a chart from BigQuery data using `Matplotlib` and saves it to
+   a Cloud Storage bucket.
+4. `Notifier` is another Python service that receives the
+   `com.google.cloud.storage.object.finalize` event from the bucket via a
+   `CloudStorageSource` and sends an email notification to users using SendGrid.
+
+### Test the pipeline
+
+`CloudSchedulerSource` creates `CloudScheduler` Jobs under the covers:
+
+```bash
+gcloud scheduler jobs list
+
+ID                                                  LOCATION      SCHEDULE (TZ)          TARGET_TYPE  STATE
+cre-scheduler-2bcb33d8-3165-4eca-9428-feb99bc320e2  europe-west1  0 16 * * * (UTC)       Pub/Sub      ENABLED
+cre-scheduler-714c0b82-c441-42f4-8f99-0e2eac9a5869  europe-west1  0 17 * * * (UTC)       Pub/Sub      ENABLED
+```
+
+And you can trigger them on demand:
+
+```bash
+gcloud scheduler jobs run cre-scheduler-2bcb33d8-3165-4eca-9428-feb99bc320e2
+```
+
+You should get an email with with a chart similar to this in a few minutes:
+
+![Chart - United Kingdom](https://atamel.dev/img/2020/chart-unitedkingdom.png)
+
+## Knative Goodness
+
+In these pipelines, I relied on a few Knative components that greatly simplified
+my development. More specifially:
+
+1. [Eventing Sources](https://knative.dev/docs/eventing/sources/) allow you to
+   read external events in your cluster and [Knative-GCP
+   Sources](https://github.com/google/knative-gcp#knative-gcp-sources) provide a
+   number of eventing sources ready to read events from various Google Cloud
+   sources.
+2. [Broker and Trigger](https://knative.dev/docs/eventing/broker/) provide an
+   eventing backbone where right events are delivered to right event consumers
+   without producers or consumers having to know about how the events are routed.
```

Comment:
```suggestion
2. [Broker and Trigger](https://knative.dev/docs/eventing/broker/) provide 
   event delivery without producers or consumers needing to know about how the events are 
   routed.
```

Reply:
I made this change manually as diffs were getting confused